### ISSUE TITLE
Deny requests not sent via Cloudflare

### ISSUE THREAD

**Issue by tomodwyer** [opensafely-core/opencodelists#2145]
We should set up nginx to deny any requests from users that try to bypass Cloudflare.

An [example of this can be seen of this work can be seen here](https://serverfault.com/a/887203), and [Cloudflare's list of IPs is available here](https://www.cloudflare.com/en-gb/ips/).

**Comment by iaindillingham** [comment]
Time box to half a day and speak to Simon beforehand.

**Comment by bloodearnest** [comment]
Currently, we proxy access to opencodeslists from the secure backends in order to retrieve the latest VMD mappings, which are needed at job run time.

I think this proxying does go from dokku4 via cloudflare to opencodelists, but we should check that it works with this in place: The proxied url is here:

https://github.com/opensafely-core/cohort-extractor/blob/main/cohortextractor/update_vmp_mapping.py#L40

Also, it would be good to preserve access from more than just cloudflare IPs

e.g. from localhost, and both of dokku3's IPs, and possibly from freshstatus as well, depending on how those checks are configured. This allows healthchecks and other things to still work even if cloudflare is down (unlikley) or misconfigured (more likely).

All in all, I am -0 on this idea.  I am not sure that it provides very much additional security, where as I am sure it provides some additional footguns. YMMV.

**Comment by evansd** [comment]
We ended up implementing this for OpenPrescribing as people were doing an end-run around Cloudflare in order to scrap us without rate limiting which resulted in a (thoughtless but non-malicious) DoS attack. It seems like a sensible move to me.

I'm not quite sure I follow Simon's objections here. FreshStatus should go via Cloudflare as that how users interact with site and if anything in that chain isn't working then the site is down.

We should definitely allow requests from localhost for the internal container healthchecks, but that pretty standard practice I think.

I'm not sure why the other Dokkus need direct access.

**Comment by bloodearnest** [comment]
> We ended up implementing this for OpenPrescribing as people were doing an end-run around Cloudflare in order to scrap us without rate limiting which resulted in a (thoughtless but non-malicious) DoS attack. It seems like a sensible move to me.
> 
> I'm not quite sure I follow Simon's objections here.

I'm not sure the juice is worth the squeeze is all.

> FreshStatus should go via Cloudflare as that how users interact with site and if anything in that chain isn't working then the site is down.

I prefer having direct access for check, *as well as* e2e. If one goes down and the other doesn't it helps quickly diagnose the problem.  But it was given as a suggestion to think about, that is all.

> We should definitely allow requests from localhost for the internal container healthchecks, but that pretty standard practice I think.

It is, but has not been mentioned yet.

> I'm not sure why the other Dokkus need direct access.

Me neither. I didn't suggest that? Dokku3 has two ips...

**Comment by tomodwyer** [comment]
To add some context to this, when Iain and Katie were investigating #2154, #2148, and #2156 – the question came up around where the security settings are for users accessing OpenCodelists. Are rules (for example) around HTTP host headers stored in Django, Nginx, or Cloudflare? And the answer was: all of them, there is no single source for the HTTP header security rules for OpenCodelists.

Some of the team were unaware of what Cloudflare does for our domains (other than basic DNS services), and how it proxies our requests and stops various security issues from occurring as it blocks them before they hit our servers.

The discussion then because a question of, if we are using Cloudflare to protect the domain, should all requests go via Cloudflare so that we can trust their security settings to do what we expect? If not, then should we remove Cloudflare from the equation entirely? A malicious user can trivially bypass Cloudflare by simply editing their hosts file.

---

In response to some of the things raised so far:

> I think this proxying does go from dokku4 via cloudflare to opencodelists, but we should check that it works with this in place

I was not aware of that, I will flag with the team.

> Also, it would be good to preserve access from more than just cloudflare IPs
>
> e.g. from localhost, and both of dokku3's IPs, and possibly from freshstatus as well, depending on how those checks are configured

- As per #2154, the web health check from `CHECKS` bypasses Cloudflare and nginx already.
- We can add the dokku3 IP addresses to the allow list
- As far as I'm aware, Freshstatus doesn't provide us with IP addresses they use for checking uptime, so that would continue as it currently is, going via Cloudflare

**Comment by lucyb** [comment]
Would Digital Ocean's Firewall help here? Rather than configuring the rules in Nginx on Dokku3, we could configure them on DO's Firewall instead. This might make it easier to manage and maybe help with some of Simon's concerns?

In terms of why we want to do this.. using Cloudflare will provide us with bad bot protection and potentially other things at the application layer, which will be really helpful. I think the security benefits will be limited because the server IP addresses will still be externally accessible, however, that's not the main motivation for doing the work.

The question for me is whether we want to delay doing this until after we've added DO's firewall?

**Comment by bloodearnest** [comment]
> Would Digital Ocean's Firewall help here? Rather than configuring the rules in Nginx on Dokku3, we could configure them on DO's Firewall instead. This might make it easier to manage and maybe help with some of Simon's concerns?

I don't think so, as dokku3 hosts multiple sites, not just opencodelists, and they all share port 443.
Unless DO firewall is snazzy and does SNI host based blocking? I doubt it, as Encrypted SNI is becoming the norm, and advocated by cloudflare: https://www.cloudflare.com/en-gb/learning/ssl/what-is-encrypted-sni/

AIUI, this currently needs to be done at the nginx vhost level, e.g. by dropping a file into `~dokku/opencodelists/nginx.conf.d/` on dokku3 with an allowlist of IPs.
